File: solutions/linked-list/143-reorder-list.py
I'll help you implement the Reorder List solution in Python. This solution will reorder a linked list by connecting the first element with the last, second with second to last, and so on.

```python
# Definition for singly-linked list node
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Reorders the linked list in-place following the pattern:
        L0 → Ln → L1 → Ln-1 → L2 → Ln-2 → ...
        
        Args:
            head: Head of the linked list
            
        Returns:
            None (modifies the list in-place)
        """
        if not head or not head.next:
            return
        
        # Step 1: Find the middle of the linked list
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        # Step 2: Reverse the second half of the linked list
        second = slow.next
        slow.next = None  # Break the list into two parts
        prev = None
        
        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp
        
        # Step 3: Merge the two halves
        first = head
        second = prev
        
        while second:
            temp1 = first.next
            temp2 = second.next
            
            first.next = second
            second.next = temp1
            
            first = temp1
            second = temp2

def print_list(head: ListNode) -> None:
    """Helper function to print the linked list"""
    current = head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")

# Example usage and test cases
def test_reorder_list():
    # Test Case 1: Normal case
    # Input: 1->2->3->4
    # Output: 1->4->2->3
    head1 = ListNode(1)
    head1.next = ListNode(2)
    head1.next.next = ListNode(3)
    head1.next.next.next = ListNode(4)
    
    print("Test Case 1:")
    print("Before reordering:", end=" ")
    print_list(head1)
    Solution().reorderList(head1)
    print("After reordering:", end=" ")
    print_list(head1)
    
    # Test Case 2: Odd length list
    # Input: 1->2->3->4->5
    # Output: 1->5->2->4->3
    head2 = ListNode(1)
    head2.next = ListNode(2)
    head2.next.next = ListNode(3)
    head2.next.next.next = ListNode(4)
    head2.next.next.next.next = ListNode(5)
    
    print("\nTest Case 2:")
    print("Before reordering:", end=" ")
    print_list(head2)
    Solution().reorderList(head2)
    print("After reordering:", end=" ")
    print_list(head2)
    
    # Test Case 3: Single node
    head3 = ListNode(1)
    print("\nTest Case 3:")
    print("Before reordering:", end=" ")
    print_list(head3)
    Solution().reorderList(head3)
    print("After reordering:", end=" ")
    print_list(head3)

if __name__ == "__main__":
    test_reorder_list()
```

This implementation includes:

1. A `ListNode` class for the linked list structure
2. The main `reorderList` solution that works in three steps:
   - Finding the middle of the list
   - Reversing the second half
   - Merging the two halves
3. Helper function to print the linked list
4. Test cases with different scenarios
5. Clear comments explaining the implementation
6. Proper handling of edge cases

The solution follows these steps:
1. First, it finds the middle of the list using slow and fast pointers
2. Then it reverses the second half of the list
3. Finally, it merges the first half with the reversed second half

The time complexity is O(n) where n is the number of nodes in the list, and the space complexity is O(1) as it modifies the list in-place.

The code handles various edge cases including:
- Empty list
- Single node list
- Even and odd length lists

When you run this code, it will demonstrate the reordering with different test cases and print the results.

File: solutions/linked-list/143-reorder-list.js
I'll help you create a complete JavaScript implementation for the Reorder List problem. I'll write the solution with clear comments and proper structure.

```javascript
/**
 * Definition for singly-linked list node
 */
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

/**
 * Reorders a linked list in the following pattern:
 * L0 → Ln → L1 → Ln-1 → L2 → Ln-2 → ...
 * 
 * @param {ListNode} head - Head of the linked list
 * @return {void} Modifies the list in place
 */
function reorderList(head) {
    // Handle edge cases
    if (!head || !head.next || !head.next.next) {
        return;
    }

    // Step 1: Find the middle of the linked list
    let slow = head;
    let fast = head;
    while (fast.next && fast.next.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    // Step 2: Reverse the second half of the list
    let secondHalf = reverseList(slow.next);
    slow.next = null; // Break the list into two halves
    
    // Step 3: Merge the two halves
    mergeAlternating(head, secondHalf);
}

/**
 * Reverses a linked list
 * @param {ListNode} head - Head of the list to reverse
 * @return {ListNode} Head of the reversed list
 */
function reverseList(head) {
    let prev = null;
    let current = head;
    
    while (current) {
        let nextTemp = current.next;
        current.next = prev;
        prev = current;
        current = nextTemp;
    }
    
    return prev;
}

/**
 * Merges two lists by alternating nodes
 * @param {ListNode} list1 - First list
 * @param {ListNode} list2 - Second list
 */
function mergeAlternating(list1, list2) {
    while (list1 && list2) {
        let temp1 = list1.next;
        let temp2 = list2.next;
        
        list1.next = list2;
        list2.next = temp1;
        
        list1 = temp1;
        list2 = temp2;
    }
}

/**
 * Helper function to create a linked list from an array
 * @param {Array} arr - Array of values
 * @return {ListNode} Head of the created linked list
 */
function createLinkedList(arr) {
    if (!arr.length) return null;
    
    const head = new ListNode(arr[0]);
    let current = head;
    
    for (let i = 1; i < arr.length; i++) {
        current.next = new ListNode(arr[i]);
        current = current.next;
    }
    
    return head;
}

/**
 * Helper function to convert linked list to array for testing
 * @param {ListNode} head - Head of the linked list
 * @return {Array} Array representation of the linked list
 */
function linkedListToArray(head) {
    const result = [];
    let current = head;
    
    while (current) {
        result.push(current.val);
        current = current.next;
    }
    
    return result;
}

// Example usage and test cases
function runTests() {
    // Test Case 1: [1,2,3,4]
    let test1 = createLinkedList([1, 2, 3, 4]);
    reorderList(test1);
    console.log('Test 1:', linkedListToArray(test1)); // Expected: [1,4,2,3]

    // Test Case 2: [1,2,3,4,5]
    let test2 = createLinkedList([1, 2, 3, 4, 5]);
    reorderList(test2);
    console.log('Test 2:', linkedListToArray(test2)); // Expected: [1,5,2,4,3]

    // Test Case 3: Single node
    let test3 = createLinkedList([1]);
    reorderList(test3);
    console.log('Test 3:', linkedListToArray(test3)); // Expected: [1]

    // Test Case 4: Empty list
    let test4 = createLinkedList([]);
    reorderList(test4);
    console.log('Test 4:', linkedListToArray(test4)); // Expected: []
}

// Run the tests
runTests();

// Export the functions for external use
module.exports = {
    reorderList,
    ListNode,
    createLinkedList,
    linkedListToArray
};
```

This implementation includes:

1. A complete solution for reordering a linked list
2. Helper classes and functions for testing
3. Clear comments explaining each part of the implementation
4. Test cases demonstrating the functionality
5. Proper error handling and edge cases
6. Module exports for external use

The algorithm works in three main steps:
1. Find the middle of the linked list using slow and fast pointers
2. Reverse the second half of the list
3. Merge the two halves by alternating nodes

The time complexity is O(n) where n is the number of nodes in the list, and the space complexity is O(1) as it modifies the list in place.

The implementation includes test cases and helper functions to make it easy to verify the solution works correctly. The code follows JavaScript best practices and includes proper error handling for edge cases like empty lists or single-node lists.